File: tests/tests_js/test_tokenizer_errors.mjs
import { Tokenizer } from '../../forthic-js/tokenizer.mjs';
import { run_tests } from './utils.mjs';

// Raise exception if strings are unterminated
function test_unterminated_string() {
    let tokenizer = new Tokenizer("'Unterminated")
    try {
        tokenizer.next_token();
    }
    catch(e) {
        console.log("PASS: ", e);
        return true;
    }

    return false;
}

// Can't have an empty definition
function test_start_definition_eos() {
    let tokenizer = new Tokenizer(":")
    try {
        tokenizer.next_token();
    }
    catch(e) {
        console.log("PASS: ", e);
        return true;
    }

    return false;
}

// Can't have definition with Forthic special chars
function test_start_definition_special_chars() {

    function check_for_exception(input) {
        let res = false;
        let tokenizer = new Tokenizer(input);
        try {
            let token = tokenizer.next_token();
            console.log(token);
        }
        catch(e) {
            console.log("PASS: ", e);
            res = true;
        }
        return res;
    }

    let start_defs = [": 'HOWDY", ": HOW'DY", ": HOW[DY", ": HOW]DY", ": HOW{DY", ": HOW}DY"];

    let result = start_defs.every(input => check_for_exception(input));
    return result;
}

let tests = {
    "test_unterminated_string": test_unterminated_string,
    "test_start_definition_eos": test_start_definition_eos,
    "test_start_definition_special_chars": test_start_definition_special_chars,
}


export { tests };
// run_tests(tests);

File: forthic/tests/v2/test_tokenizer_errors.py
import unittest
from forthic.v2.tokenizer import Tokenizer, UnterminatedStringError, InvalidDefinitionError


class TestTokenizerErrors(unittest.TestCase):
    def test_unterminated_string(self):
        """Raise exception if strings are unterminated
        """
        tokenizer = Tokenizer("'Unterminated")
        self.assertRaises(UnterminatedStringError, tokenizer.next_token)

    def test_start_definition_eos(self):
        """Can't have an empty definition
        """
        tokenizer = Tokenizer(":")
        self.assertRaises(InvalidDefinitionError, tokenizer.next_token)

    def test_start_definition_special_chars(self):
        """Can't have definition with Forthic special chars
        """
        def check_for_exception(input):
            tokenizer = Tokenizer(input)
            self.assertRaises(InvalidDefinitionError, tokenizer.next_token)

        invalid_start_defs = [": 'HOWDY", ": HOW'DY", ": HOW[DY", ": HOW]DY", ": HOW{DY", ": HOW}DY"]

        for forthic in invalid_start_defs:
            check_for_exception(forthic)


if __name__ == '__main__':
    unittest.main()
